### DISCUSSION TITLE
Is it possible to supply a list of subtypes instead of just one?

### DISCUSSION THREAD

**Discussion by demlik** [instancio/instancio#457] [Q&A]
Hello,

When testing, I would like to be able to supply a list of subtypes and instancio should create one of each concrete subtypes when it encounters a collection

How might I do this?

**Comment by armandino** [comment]
@demlik, thanks for the question. Are you using JUnit 5? If so, you could use a `ParameterizedTest`:

```java
class ListHolder {
    List<String> list;
}

@ValueSource(classes = {ArrayList.class, LinkedList.class})
@ParameterizedTest
void subtypes(Class<?> listSubtype) {
    ListHolder result = Instancio.of(ListHolder.class)
            .subtype(all(List.class), listSubtype)
            .create();
}
```

**Reply by demlik** [reply]
Thanks for the suggestion, I am using JUnit4.

Let me give a more concrete example:

class Zoo {
 List<Animal animals;
}

interface Animal {}
class Lion implements Animal {}
class Tiger implements Animal {} 

How can I create an instance of Zoo containing both Lion and Tiger?

**Comment by armandino** [comment]
In this case, you could select the animal type randomly:

```java
@Test
void example() {
    Class<?> animalType = Gen.oneOf(Lion.class, Tiger.class).get();

    Zoo zoo = Instancio.of(Zoo.class)
        .subtype(all(Animal.class), animalType)
        .create();
}
```
If random subtype doesn't work, then refactoring the test is another option

```java
@Test
void example() {
    assertZooWithAnimal(Lion.class);
    assertZooWithAnimal(Tiger.class);
}

void assertZooWithAnimal(Class<?> animalType) {
    Zoo zoo = Instancio.of(Zoo.class)
            .subtype(all(Animal.class), animalType)
            .create();
    // ...
}
```

JUnit 4 also supports parameterized tests, but I don't recall the syntax (it is more verbose than JUnit 5)/

**Reply by demlik** [reply]
Is there no way to populate the list with both entries at the same time?

**Reply by armandino** [reply]
That's an interesting idea. Currently, the API doesn't support such a use case directly, but it can be done using a custom generator to create the elements:

```java
@Test
void randomElementTypes() {
    Class<?>[] subtypes = {Lion.class, Tiger.class};
    
    Generator<Animal> animalGenerator = random -> Instancio.of(Animal.class)
            .subtype(all(Animal.class), random.oneOf(subtypes))
            .create();

    Zoo zoo = Instancio.of(Zoo.class)
            .supply(all(Animal.class), animalGenerator)
            .create();
}
```

I think it could be a useful feature to support this via a top-level API. I'll need to think about it some more. Thank you for bringing this up :)

**Reply by demlik** [reply]
Above solution is workable for me - thank you for the fast response! It would be nice to have top level support :)

**Reply by armandino** [reply]
@demlik any time :) I created #458 to investigate this further.

Let me know if you have any other questions.

**Reply by armandino** [reply]
@demlik unfortunately supporting this feature via top-level API is not feasible. I added some background information in #458.
